### PR TITLE
Long running tasks (#442)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
```
from time import sleep
from src.taipy.gui import Gui, State, invoke_long_running, notify
import threading as thr
import typing as t

status = ""


def heavy_function(*args):
    print(f"{thr.current_thread().name}: heavy_function.started {len(args)} {args}")
    sleep(5)
    if (len(args) > 4):
        raise Exception("Heavy function Exception")
    sleep(10)
    print(f"{thr.current_thread().name}: heavy_function.ended {args}")


def heavy_function_status(state: State, status: bool | int, p1, p2, p3):
    # Now the State can be used to update the user interface
    print(f"{thr.current_thread().name}: heavy_function_status {status} {p1} {p2} {p3}")
    ntf_status = ("Done" if status and isinstance(status, bool) else f"{status}")
    state.status = "Done" if status and isinstance(status, bool) else ("." * (status + 1))
    notify(state, "info", f"The heavy task is {ntf_status}!")

def on_exception(state: State, function_name: str, e: Exception):
    state.status = f"Exception on {function_name} {e}"
    notify(state, "Error", state.status)

def on_action(state: State, *action_args):
    notify(state, "info", "Heavy task started...")
    state.status = "Started "
    # Execute the heavy task, in a new thread
    invoke_long_running(state, heavy_function, ["HF.Param 1", 2, True, [1, 2, 3]],
                        heavy_function_status, ["HFD.Param1", 3, False], 1)
    # invoke_long_running(state, heavy_function)


Gui("<|{status}|><|Start long running function|button|>").run(use_reloader=False)

```

### CODE REVIEW

File: src/taipy/gui/__init__.py [hunk 1]
```diff
@@ -72,10 +72,11 @@
 from .gui import Gui
 from .gui_actions import (
     download,
-    get_state_id,
     get_module_name_from_state,
+    get_state_id,
     hold_control,
     invoke_callback,
+    invoke_long_running,
```
Comment (FabienLelaquais) [inline src/taipy/gui/__init__.py]: I would prefer an `invoke_long_callback` (to match the one above).

File: src/taipy/gui/gui.py [hunk 2]
```diff
@@ -914,7 +914,8 @@ def _call_user_callback(self, state_id: t.Optional[str], user_callback: t.Callab
                 self.__set_client_id_in_context(state_id)
                 return self._call_function_with_state(user_callback, args)
         except Exception as e:
-            warnings.warn(f"Exception raised in invoke_callback: '{user_callback.__name__}':\n{e}")
+            if not self._call_on_exception(user_callback.__name__, e):
+                warnings.warn(f"invoke_callback: Exception raised in function : '{user_callback.__name__}':\n{e}")
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui.py]: Error message should be
f"invoke_callback: Exception raised in function '{user_callback.\_\_name\_\_}'.\n{e}")

To match the one in on_init

File: src/taipy/gui/gui_actions.py [hunk 3]
```diff
@@ -195,3 +196,74 @@ def invoke_callback(gui: Gui, state_id: str, callback: t.Callable, args: t.Union
 def invoke_state_callback(gui: Gui, state_id: str, callback: t.Callable, args: t.Union[t.Tuple, t.List]) -> t.Any:
     warnings.warn("'invoke_state_callback()' was deprecated in Taipy GUI 2.0. Use 'invoke_callback()' instead.")
     return invoke_callback(gui, state_id, callback, args)
+
+
+def invoke_long_running(
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui_actions.py]: invoke_long_callback

File: src/taipy/gui/gui_actions.py [hunk 4]
```diff
@@ -195,3 +196,74 @@ def invoke_callback(gui: Gui, state_id: str, callback: t.Callable, args: t.Union
...
+    state: State,
+    user_function: t.Callable,
+    user_function_args: t.Union[t.Tuple, t.List] = [],
+    user_status_function: t.Optional[t.Callable] = None,
+    user_status_function_args: t.Union[t.Tuple, t.List] = [],
+    delay=0,
+):
+    """Invoke a long running user callback (uses Threads).
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui_actions.py:208]: I'll update the RefMan when I do the UserMan part.
This is kind of obsolete now...

File: src/taipy/gui/gui_actions.py [hunk 5]
```diff
@@ -195,3 +196,74 @@ def invoke_callback(gui: Gui, state_id: str, callback: t.Callable, args: t.Union
...
+                if active_thread:
+                    callback_on_status(count)
+                    threading.Timer(delay, thread_status, (name, delay, count + 1)).start()
+
+            thread = threading.Thread(target=user_function_in_thread, args=user_function_args)
+            thread.start()
+            if delay and isinstance(delay, int) and callable(user_status_function):
+                thread_status(thread.name, delay, 0)
+    else:
+        warnings.warn("'invoke_long_running()' must be called in the context of a callback")
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui_actions.py]: -> callback

File: src/taipy/gui/gui_actions.py [hunk 6]
```diff
@@ -195,3 +196,74 @@ def invoke_callback(gui: Gui, state_id: str, callback: t.Callable, args: t.Union
...
+            When set to 0, the user_status_fucntion is called only when the user_function is terminated.
+    """
+    if state and isinstance(state._gui, Gui):
+        state_id = get_state_id(state)
+        if isinstance(state_id, str):
+            this_gui = state._gui
+
+            def callback_on_exception(state: State, function_name: str, e: Exception):
+                if not this_gui._call_on_exception(function_name, e):
+                    warnings.warn(f"invoke_long_running: Exception raised in function {function_name}.\n{e}")
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui_actions.py]: -> callback

File: src/taipy/gui/gui_actions.py [hunk 7]
```diff
@@ -195,3 +196,74 @@ def invoke_callback(gui: Gui, state_id: str, callback: t.Callable, args: t.Union
...
     return invoke_callback(gui, state_id, callback, args)
+
+
+def invoke_long_running(
+    state: State,
+    user_function: t.Callable,
+    user_function_args: t.Union[t.Tuple, t.List] = [],
+    user_status_function: t.Optional[t.Callable] = None,
+    user_status_function_args: t.Union[t.Tuple, t.List] = [],
+    delay=0,
```
Comment (FabienLelaquais) [inline src/taipy/gui/gui_actions.py]: Rename to 'period'
Unit should be milliseconds for homogeneity, with a minimal value of 500 (or whatever)

**Review by FabienLelaquais — APPROVED** [review]
I love it!

File: src/taipy/gui/gui_actions.py [hunk 8]
```diff
@@ -195,3 +196,74 @@ def invoke_callback(gui: Gui, state_id: str, callback: t.Callable, args: t.Union
...
+        user_function_args (Optional[List|Tuple]): The remaining arguments, as a List.
+        user_status_function (Optional(Callable[[State^, bool, ...], None])): The optional user-defined status function that is invoked at the end (of and possibly during) the runtime of the user_function.<br/>
+            The first parameter of this function **must** be a `State^`. The second parameter of this function **must** be a Union[`bool`, `int`] that will indicate a progression (`int`) or the final status (`bool`)
+        user_status_function_args (Optional[List|Tuple]): The remaining arguments ot the user status function, as a List.
+        period (int): The interval in milli-seconds at which the user_status_function is called (default value is 0).</br>
+            When set to a value < 500, the user_status_function is called only when the user_function is terminated.
+    """
+    if state and isinstance(state._gui, Gui):
+        state_id = get_state_id(state)
+        if isinstance(state_id, str):
```
Comment (dinhlongviolin1) [inline src/taipy/gui/gui_actions.py]: if not isinstance(state_id, str):
    return
..then the rest

File: src/taipy/gui/gui_actions.py [hunk 9]
```diff
@@ -195,3 +196,74 @@ def invoke_callback(gui: Gui, state_id: str, callback: t.Callable, args: t.Union
...
+        state (State^): The `State^` instance, which is an argument of the callback function.
+        user_function (Callable[[...], None]): The function that will be run independently of taipy gui (cannot use `State^`).
+        user_function_args (Optional[List|Tuple]): The remaining arguments, as a List.
+        user_status_function (Optional(Callable[[State^, bool, ...], None])): The optional user-defined status function that is invoked at the end (of and possibly during) the runtime of the user_function.<br/>
+            The first parameter of this function **must** be a `State^`. The second parameter of this function **must** be a Union[`bool`, `int`] that will indicate a progression (`int`) or the final status (`bool`)
+        user_status_function_args (Optional[List|Tuple]): The remaining arguments ot the user status function, as a List.
+        period (int): The interval in milli-seconds at which the user_status_function is called (default value is 0).</br>
+            When set to a value < 500, the user_status_function is called only when the user_function is terminated.
+    """
+    if state and isinstance(state._gui, Gui):
```
Comment (dinhlongviolin1) [inline src/taipy/gui/gui_actions.py]: if not state or not isinstance(state._gui, Gui):
   warnings.warn("'invoke_long_callback()' must be called in the context of a callback.")
   return

You get what I mean :D

**Review by dinhlongviolin1 — APPROVED** [review]
NICEEEEEEEEEE

File: src/taipy/gui/gui_actions.py (hunk 9 above)
Comment (FredLL-Avaiga) [reply src/taipy/gui/gui_actions.py]: Less visible complexity...
Yes...

**Pull request merged**